### PR TITLE
Add build setting for leak checking.

### PR DESCRIPTION
**Pull request by Teemperor**
Previously these build options where hidden behind a define and the
only way to test them is to manually add this to the CXX/CC flags
by hand.

This patch turns this into a proper (but experimental) build setting
to better document the effort to have proper memory deallocation in
ROOT. It also fixes some of the broken code that was hidden behind
this define.

### CODE REVIEW

**Comment by Teemperor** [comment]
Let's see if this works.

File: core/meta/src/TClass.cxx [hunk 1]
```diff
@@ -371,15 +371,13 @@ namespace ROOT {
...
-#ifdef R__COMPLETE_MEM_TERMINATION
       TMapTypeToTClass() {
          TIter next(&fMap);
          TObjString *key;
          while((key = (TObjString*)next())) {
             delete key;
          }
-      }
-#endif
+      }=
```
Comment (amadio) [inline core/meta/src/TClass.cxx]: This line looks like it was left over during editing.

Comment (Teemperor) [inline core/meta/src/TClass.cxx]: Thanks, fixed!

**Comment by Teemperor** [comment]
Ok, the code is broken beyond repair, let's just kill it with fire.

I updated the PR to remove the code, the previous failed builds were trying to see if anything behind this define is still serving some purpose.

**Comment by vgvassilev** [comment]
@Teemperor IIRC @pcanal was using this switch when he was working on ROOT destruction. We had some issues when deleting the memory on exit. Let's wait for his opinion before removing this.

It seems challenging to get ROOT's destruction order right but if we are still looking for such PR :)

**Comment by pcanal** [comment]
> Ok, the code is broken beyond repair, let's just kill it with fire.

Yes, getting the code to work properly is 'hard' and thus the work has not been completely.  If you have analyzed each of those lines and deemed that indeed it is not semantically possible to cleanups those things at the end of the process, then please replace the (unactivated) deletion code with a clear explanation of why it is impossible to delete (and saying 'it is hard' or 'broken beyond repair' is not quite enough of an explanation :) :) ).

So unless, somebody actually investigate each of those lines, please leave them as a guide of what still needs to be investigated.

Thanks.

**Comment by Teemperor** [comment]
If we leave it, we should add least have a Jenkins build for it and some documented build setting that explains this without having to search in the git log. Otherwise people just add more and more typos between those ifdefs, remove the things the code in there references, or fix the memory leaks in another way, causing this code to be just misleading at one point in the future.

**Comment by vgvassilev** [comment]
@pcanal, I'd like to proceed with this PR. What about merging it and immediately open another PR with these changes, this way somebody could work on this in the context of the PR. This would be a good middle ground between your and @Teemperor's point.

**Comment by pcanal** [comment]
Unless somebody is going to work on it 'immediately', this plan would, in my opinion, lead to the PR being eventually 'forgotten' about, authors then re-adding the removed code piece-meal (and without the insight of the previous tries).
If there is typos and removal, they will be discovered (and fixed) by whomever is courageous enough to try actually turning it on and fixing the underlying issues.

**Comment by vgvassilev** [comment]
I agree but the pro would be that we will have less 'broken'/half-baked code in ROOT while at the same time we keep track of it in a very public way.

**Comment by Teemperor** [comment]
@phsft-bot build with flags -Dmemory_termination=ON

**Comment by pcanal** [comment]
> I agree but the pro would be that we will have less 'broken'/half-baked code in ROOT while at the same time we keep track of it in a very public way.

@vgvassilev I respectfully disagree.  Unless somebody is actively maintaining this branch (and that time might actually be better used to actively resolve the issues), the branch will soon become unmergeable (without 'work') and thus be almost useless.  Furthermore, if the code is no longer in the actual source file, the branch will soon(ish) be forgotten and will almost certainly not be checked against when modifying the relevant part of the code (and would could remember all the parts of the code covered).   So I would rather have in the source code, code fragments that might possibly get updated (when people use search/replace, etc.) rather than a branch that will most likely (again, unless somebody is actively working on it) be never updated ....

File: core/base/src/TROOT.cxx [hunk 2]
```diff
@@ -930,7 +930,7 @@ TROOT::~TROOT()
       SafeDelete(fGlobals);
       if (fGlobalFunctions) fGlobalFunctions->Delete();
       SafeDelete(fGlobalFunctions);
-      fEnums->Delete();
+      fEnums.load()->Delete();
```
Comment (pcanal) [inline core/base/src/TROOT.cxx]: Isn't the load implement in the ->?

Comment (Teemperor) [inline core/base/src/TROOT.cxx]: [cppreference says no](http://en.cppreference.com/w/cpp/atomic/atomic) :(

**Comment by Teemperor** [comment]
@peremato Any objection against a Jenkins build that tests this build option from time to time? I would like @pcanal work to converge with master at some point and this would keep track of the progress (or any further regressions).

File: core/base/src/TROOT.cxx (hunk 2 above)
Comment (pcanal) [inline core/base/src/TROOT.cxx]: My bad, I was off by one level of indirection :)
And indeed, elsewhere in the code I have used
```
(*fEnums).Delete();
```
thanks.

**Comment by pcanal** [comment]
@myself @vgvassilev @Teemperor  I just realize that the nature of the PR completely changed (to match the direction I was pushing) so at this point (i.e. adding cmake flag to allow testing) I think this is good to go.  thanks.

**Comment by Teemperor** [comment]
@phsft-bot build

**Comment by Teemperor** [comment]
@phsft-bot build with flags -Dmemory_termination=ON

**Comment by Teemperor** [comment]
@phsft-bot build with flags -Dmemory_termination=ON

The few changes I did should make rootcling pass and we can run tests. Let's see if it works.

**Comment by Teemperor** [comment]
@phsft-bot build

**Comment by Teemperor** [comment]
Ok, this doesn't break master and the clang-format changes would actually introducing inconsistencies here, so I think this is good to go. I'll merge so that @martinmine can setup the Jenkins builds and we can track progress of this.

File: core/base/src/TROOT.cxx [hunk 3]
```diff
@@ -930,8 +933,10 @@ TROOT::~TROOT()
       SafeDelete(fGlobals);
       if (fGlobalFunctions) fGlobalFunctions->Delete();
       SafeDelete(fGlobalFunctions);
-      fEnums->Delete();
-      fClasses->Delete();    SafeDelete(fClasses);     // TClass'es must be deleted last
+      fEnums.load()->Delete();
+
+      // FIXME: Causes segfault in rootcling, debug and uncomment
+      // fClasses->Delete();    SafeDelete(fClasses);     // TClass'es must be deleted last
```
Comment (pcanal) [inline core/base/src/TROOT.cxx]: FYI.  The  intent was to move the 'good delete' out of the #ifdef and keep the bad one hidden behind it ....

**Comment by pcanal** [comment]
> Ok, the code is broken beyond repair, let's just kill it with fire.

Yes, getting the code to work properly is 'hard' and thus the work has not been completely.  If you have analyzed each of those lines and deemed that indeed it is not semantically possible to cleanups those things at the end of the process, then please replace the (unactivated) deletion code with a clear explanation of why it is impossible to delete (and saying 'it is hard' or 'broken beyond repair' is not quite enough of an explanation :) :) ).

So unless, somebody actually investigate each of those lines, please leave them as a guide of what still needs to be investigated.

Thanks.

**Pull request merged**